### PR TITLE
BlockId removal: refactor: StorageProvider

### PR DESCRIPTION
**Pull request by michalkucharczyk**
It changes the arguments of `Backend::StorageProvider` trait from:
block: `BlockId<Block>` to: hash: `&Block::Hash`

This PR is part of BlockId::Number refactoring analysis (paritytech/substrate#11292)

polkadot companion: paritytech/polkadot#6160
cumulus companion: paritytech/cumulus#1770

### CODE REVIEW

File: client/cli/src/commands/export_state_cmd.rs [hunk 1]
```diff
@@ -57,7 +57,7 @@ impl ExportStateCmd {
 	) -> error::Result<()>
 	where
 		B: BlockT,
-		C: UsageProvider<B> + StorageProvider<B, BA>,
+		C: UsageProvider<B> + StorageProvider<B, BA> + HeaderBackend<B>,
```
Comment (bkchr) [inline client/cli/src/commands/export_state_cmd.rs:60]:
```suggestion
		C: UsageProvider<B> + StorageProvider<B, BA>,
```

File: client/finality-grandpa/src/import.rs [hunk 2]
```diff
@@ -431,6 +431,12 @@ where
 				e
 			))
 		})?;
+
+		let hash = self
+			.inner
+			.expect_block_hash_from_id(id)
+			.map_err(|e| ConsensusError::Other(e.into()))?;
```
Comment (bkchr) [inline client/finality-grandpa/src/import.rs:438]:
```suggestion
```

Please make this function take `Block::Hash` directly.

File: client/service/test/src/client/mod.rs [hunk 3]
```diff
@@ -1602,12 +1602,14 @@ fn storage_keys_iter_prefix_and_start_key_works() {
 		.add_extra_child_storage(&child_info, b"third".to_vec(), vec![0u8; 32])
 		.build();
 
+	let block_hash = client.block_hash_from_id(&BlockId::Number(0)).unwrap().unwrap();
```
Comment (bkchr) [inline client/service/test/src/client/mod.rs:1605]:
```suggestion
	let block_hash = client.info().best_hash;
```

File: client/service/test/src/client/mod.rs [hunk 4]
```diff
@@ -1666,10 +1668,12 @@ fn storage_keys_iter_prefix_and_start_key_works() {
 fn storage_keys_iter_works() {
 	let client = substrate_test_runtime_client::new();
 
+	let block_hash = client.block_hash_from_id(&BlockId::Number(0)).unwrap().unwrap();
```
Comment (bkchr) [inline client/service/test/src/client/mod.rs:1671]:
```suggestion
	let block_hash = client.info().best_hash;
```

**Review by bkchr — APPROVED** [review]

**Review by dmitry-markin — APPROVED** [review]
LGTM!

Is it possible to move block hash from `BlockId` lookup one level up out of the functions affected, so that there is no `HeaderBackend` constraint added? Or is it planned for the next iterations of the refactoring?

**Review by altonen — APPROVED** [review]

File: client/cli/src/commands/export_state_cmd.rs (hunk 1 above)
Comment (bkchr) [reply client/cli/src/commands/export_state_cmd.rs:60]: Wrong comment.

**Review by skunert — APPROVED** [review]

File: client/service/src/chain_ops/export_raw_state.rs [hunk 5]
```diff
@@ -30,14 +30,15 @@ pub fn export_raw_state<B, BA, C>(
...
 ) -> Result<Storage, Error>
 where
-	C: UsageProvider<B> + StorageProvider<B, BA>,
+	C: UsageProvider<B> + StorageProvider<B, BA> + HeaderBackend<B>,
 	B: BlockT,
 	BA: sc_client_api::backend::Backend<B>,
 {
 	let block = block.unwrap_or_else(|| BlockId::Hash(client.usage_info().chain.best_hash));
 
+	let hash = client.expect_block_hash_from_id(&block)?;
```
[the remark this answers is not in the document]
Comment (michalkucharczyk) [reply client/service/src/chain_ops/export_raw_state.rs:39]: `export_raw_state` is CLI utility, so the conversion seems to be fine here. I would leave it as is.

File: client/cli/src/commands/export_state_cmd.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (michalkucharczyk) [reply client/cli/src/commands/export_state_cmd.rs:60]: Yes, as we pass `client` to `sc_service::chain_ops::export_raw_state` where the conversion is made.

File: client/finality-grandpa/src/import.rs (hunk 2 above)
Comment (michalkucharczyk) [reply client/finality-grandpa/src/import.rs]: done in [11c7013](https://github.com/paritytech/substrate/pull/12510/commits/11c7013119e839560327b126f2a64be00edc3cba)

File: client/service/src/chain_ops/export_raw_state.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (michalkucharczyk) [reply client/service/src/chain_ops/export_raw_state.rs]: idk, i reworked that, don't know if that is better: [f0a12a8](https://github.com/paritytech/substrate/pull/12510/commits/f0a12a8ad7d5a3a8248f24fc28dc2d98b852b86e)
The bound is gone, but on the other hand this utility may require conversion in future.

**Review by bkchr — APPROVED** [review]

**Comment by michalkucharczyk** [comment]
bot merge

**Pull request merged**